### ISSUE TITLE
No way to sample from multisampled textures in impellerc shaders

### ISSUE THREAD

**Issue by bdero** [flutter/flutter#102525]
This is blocking support for rendering with non-pipeline blend modes (https://github.com/flutter/impeller/pull/159) because we need to bind the backdrop texture and render to an intermediate texture in this situation.

GLSL 3.3 has features that were replaced in 4+ with different primitives, but we're blocked from using newer builtins because we target GLES 2 -- and so neither can be used in impellerc shaders.

Specifically:
1. GLSL 4+ removes the `texture(sampler2DMS, vec2)` overload, and
2. the compiler blocks use of `texelFetch` because it can't be used  when we transpile to a GLSL version supported by GLES2.

And so even though we can still bind multisample textures, there's no way to actually pull data from them.

```glsl
uniform FrameInfo {
  ivec2 texture_size;
} frame_info;

uniform sampler2DMS texture_sampler_dst;
uniform sampler2D texture_sampler_src;

in vec2 v_dst_texture_coords;
in vec2 v_src_texture_coords;
out vec4 frag_color;
// Emulate SamplerAddressMode::ClampToBorder.
vec4 SampleWithBorder(sampler2D tex, vec2 uv) {
  if (uv.x > 0 && uv.y > 0 && uv.x < 1 && uv.y < 1) {
    return texture(tex, uv);
  }
  return vec4(0);
}

void main() {
  // This doesn't compile because it's not supported in GLSL 4.6 (the standard we use for authoring), although it's supported by earlier GLSL versions.
  vec4 dst = texture(texture_sampler_dst, v_dst_texture_coords);

  // This doesn't compile because it can't be used when transpiling to a GLES2 compatible GLSL version.
  ivec2 pixel_coords = ivec2(v_dst_texture_coords * frame_info.texture_size);
  vec4 dst = texelFetch(texture_sampler_dst, pixel_coords, 0);

  vec4 src = SampleWithBorder(texture_sampler_src, v_src_texture_coords);
  frag_color = src + dst - src * dst;
}
```

Possible solutions:
* It may make sense to just switch from GLSL 4.6 to GLSL 3.3 for authoring after refactoring the glyph atlas to not render using instances (I think @dnfield is currently doing this refactor?) The downside here is that we have more limitation on leveraging per-platform features.
* Alternatively, we could have the compiler set the source version to 3.3 for just GLES2, and use 4.6 for mostly everything else, enabling specialized solutions for this situation:

    ```glsl
    #ifdef IMPELLER_TARGET_OPENGLES
      vec4 dst = texture(texture_sampler_dst, v_dst_texture_coords);
    #else
      ivec2 pixel_coords = ivec2(v_dst_texture_coords * frame_info.texture_size);
      vec4 dst = texelFetch(texture_sampler_dst, pixel_coords, 0);
    #endif
    ```

**Comment by dnfield** [comment]
Instancing refactor is flutter/imepller#160

Would it make any sense to have some kind of "special" method for this, either defined in some kind of included file or in the compiler itself? E.g. make the compiler produce valid GLES shader code even when it sees `texelFetch`.

**Comment by bdero** [comment]
I made a mistake earlier when reading the 3.3 spec -- my proposed solutions don't work because there is still no multisample overload for `texture()` in 3.3 either, or in a handful of other GLSL specs I've looked at today (any?) for that matter.
Looking at the GLSL ES 1 spec, there's no concept of reading multisample textures in shader programs, and I couldn't spot any multisample extensions that add this. So I don't think we'll be able to bind multisample textures once GLES is running anyway.

With this in mind, I should actually be trying to bind the resolve texture for MSAA passes. Using the resolve texture is what I should have been doing in the first place since this is just a blend shader.
